wfsim/strax_interface.py

```diff
@@ -160,6 +177,168 @@ def instruction_from_csv(filename):
+
+
+    Usage: 
+        simulator = wfsim.McChainSimulator(st,run_id)
+        simulator.run_chain()
+        '''
+    def __init__(self,strax_context,run_id) -> None:
+        '''Sets configuration. '''
+        self.context = strax_context
+        self.file = self.context.config['fax_file']
+        self.targets = ('raw_records',)
+        self.run_id = run_id
+        if self.context.config['neutron_veto']:
+            self.targets = ('raw_records','raw_records_nv')
+
+    def _set_max(self,):
+        '''If no event_stop is specified set it to the maximum (-1=do all events)'''
+        if self.context.config['event_stop']==-1:
+            self.context.set_config(dict(event_stop=self.instructions_epix['g4id'][-1]+1))
+        else:
+            pass
+
+    def _instructions_from_epix_cut(self,):
+        '''Cut away events not within event_start & event_stop'''
+        mask=(self.instructions_epix['g4id']<self.context.config['event_stop'])&\
+             (self.instructions_epix['g4id']>=self.context.config['event_start'])
+        self.instructions_epix = self.instructions_epix[mask]
+
+
+    def instructions_from_epix(self,):
+        '''Run epix and save instructions as self.instruction.
+         For the moment we'll just process the whole file and then throw out all events we won't use
+         Epix needs to be imported in here to avoid circle imports'''
+        logging.info("Getting instructions from epix")
+        import epix
+
+        #make deepcopy cause some of the things in the setup will add unhashable stuff
+        epix_config = get_resource(self.context.config['epix_config'],fmt='json')
+        epix_config.update({'input_file':self.context.config['fax_file'],
+                            'entry_start':self.context.config['event_start'],
+                            'entry_stop':self.context.config['event_stop'],
+                            'source_rate':0})
+        epix_config = epix.run_epix.setup(epix_config)
+        self.instructions_epix=epix.run_epix.main(epix_config,return_wfsim_instructions=True)
+        self._set_max()
+        self._instructions_from_epix_cut()
+
+    def instructions_from_nveto(self,):
+        logging.info("Getting nveto instructions")
+        #TODO make config working with nveto...
+        self.context.config['nv_pmt_ce_factor'] = 1.0
+        self.instructions_nveto, self.nveto_channels, self.nveto_timings=read_optical(self.context.config)
+        
+
+    def set_timing(self,):
+        '''Set timing information in such a way to synchronize instructions for the TPC and nVeto'''
+        logging.info("Setting timings")
+        rate = self.context.config['event_rate']
+        start = self.context.config['event_start']
+        stop=self.context.config['event_stop']
+        n_events = stop-start
+
+        timings= (self.context.config['event_start']/rate + \
+                 wfsim.units.s*np.random.random(n_events)*(n_events)/rate).astype(np.int64)
```

Comment:
I think here you have to change the logic slightly as we do not write empty events/instructions. 

E.g. if you simulate 15 k events you may only see 8k in the TPC simply due to the solid angle. Hence the timing must be computed like this 
```
np.random.uniform(first_g4_id/rate, last_g4_id/rate, n_events)
```
If "rate" should represent the decay rate of the source and not the event rate in the TPC.

Reply:
There is one more thing which is not considered yet I think. You can have g4_id_tpc != g4_id_nveto. Hence you have to do something like  first_g4_id = min(g4_id_tpc, g4_id_nveto) and same for max.

Reply:
Is the first comment refering to the offset part?

For the second part. If the ids are not the same you cannot synchronize at all
Or do you mean the first g4 id in tpc in not nessecairy the first in nveto?
In that case you are right